runner/src/mill/runner/MillMain.scala

```diff
@@ -275,6 +275,28 @@ object MillMain {
     }
   }
 
+  private def parseThreadCount(threadCountRaw: Option[String]): Try[Int] = {
+    val cores = Runtime.getRuntime.availableProcessors()
+    def err(detail: String) = new Exception(
+      s"Invalid value \"${threadCountRaw.getOrElse("")}\" for flag -j/--jobs: $detail"
+    )
+    (threadCountRaw match {
+      case None => Success(cores)
+      case Some(s"${n}C") => n.toDoubleOption.map(Success(_))
+          .getOrElse(Failure(err("Failed to find a float number before \"C\".")))
+          .map(m => (m * cores).toInt)
+      case Some(s"C-${n}") => n.toIntOption.map(Success(_))
+          .getOrElse(Failure(err("Failed to find a int number after \"C-\".")))
+          .map(cores - _)
+      case Some(n) => n.toIntOption.map(Success(_))
+          .getOrElse(Failure(err("Failed to find a int number")))
+    }).flatMap {
+      case x if x < 0 => Failure(err("Calculated cores to use is a negative number."))
+      case 0 => Success(cores) // use all cores if 0
+      case x => Success(x)
+    }
+  }
```

Comment:
Few nits here:

1. Can we extract this into a helper function that we can unit test with various combinations of `detail: String` and `cores: Int`? That will give us some confidence at-a-glance that the code is doing what we expect without tedious manual testing

2. I don't think the current treatment of "0 = use all cores" makes sense for computed zeros. If someone runs `-j0.4C` on a 4 core machine, they probably don't intend to use all 4 cores. Let's treat literal 0 separately from computed numbers, and have computed numbers have a lower limit of 1 core (so `0` or `-1` becomes `1`)

3. Lets return an `Either[String, Int]` since that's all the information you need later on at the callsite